# NGS Scripts

## Description

This is a collection of scripts developed at the Oklahoma Medical Research Foundation to
assist in the analysis of so called "Next Generation Sequencing" data. The bulk are
for wrapping together a slew of other tools to build a basic sequence to VCF workflow

## Prerequisites

Most of the scripts are ruby based and there is also some perl. So hopefully you've got those.
The ruby scripts are developed under Ruby MRI 1.9.2. Additionally these scripts are like many
somewhat specific to our uses at OMRF & our systems installation. In this case a linux based
system with Sun Grid Engine, shared NFS, the [Environment Modules Project](http://modules.sourceforge.net/), 
and then some of the standard sequencing analysis packages:

1. [BWA](http://bio-bwa.sourceforge.net/)
2. [samtools](http://samtools.sourceforge.net/)
3. [picard](http://picard.sourceforge.net/)
4. [GATK](http://www.broadinstitute.org/gsa/wiki/index.php/The_Genome_Analysis_Toolkit)
5. [fastqc](http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/)
6. [tabix](http://samtools.sourceforge.net/tabix.shtml)
7. [b-tangs](https://github.com/oklasoft/b-tangs)

Additionally for some of the java based tools above, we've got wrapper shell
scripts for the commands so not to run it with the full on java -jar .... command

## Installation

Download it, set the shebang line if needed & add it to your path or put them
somewhere in your path. Hey it passes the "works for us test", so installing isn't
a high priority for me.

## Scripts

* **analyze_sequence_to_snps.rb**

  Main wrapper script, takes a config file & runs pipeline to produce VCFs of SNPs & Indels
 
* **bwa_aln_qsub_tasked.rb**

  Simple wrapper to run the aln method of bwa via a SGE task array
  
* **bwa_sampese_qsub_tasked.rb**

  Another quick wrapper script for a SGE task array, this time to bwa samse or sampe
  
* **clc_snp_reader.rb**

  Deprecated script for doing something to CLC output
  
* **clc_snp_results_to_linkage_ped.rb**

  Deprecated script again for doing something to CLC based output
  
* **convert_fastq_to_sam_qsub_tasked.rb**

  Wrapper for a SGE task array to convert the fastq input to SAM files
  
* **count_each_individual.rb**

  Quick wrapper to DepthOfCoverage a set of BAM files
  
* **fix_sam_rg_qsub_tasked.rb**

  Wrapper for SGE task array to force/fix the RG tag in the SAM file(s)
  
* **fix_vcf_generated_ped.rb**

  Correct the ped files generated by vcftools by using a full correct ped file template
  
* **gatk_depth_coverage_summation.rb**

  Combine GATK DepthOfCoverage outputs, ala count_each_individual.rb & then sum that output
  
* **make_bam_qsub_tasked.rb**

  Wrapper for SGE task array to make BAM file(s)
  
* **snip_reader.rb**

  Deprecated script again having to do something with CLC
  
* **vqsr_vcf.rb**

  Performs the GATK Variant Quality Score Recalibration workflow on a VCF
  
* **add_read_group_to_sam.pl**

  PERL script found online at [seqanswers.com](http://seqanswers.com/forums/showthread.php?t=4180) to force the RG tag, as version of BWA currently used does not add it for unaligned reads

# License

Unless otherwise noted, everything is Copyright (c) 2011 Stuart Glenn, Oklahoma Medical
Research Foundation. (OMRF). Essentially under the BSD license, see LICENSE file for
full details. 

add_read_group_to_sam.pl is excluded from that copyright & license, as it just
included here for connivence & assumed to be
publicish domain due to open sharing on the forums
